src/util/array_element_from_pointer.h

```diff
@@ -0,0 +1,32 @@
+/*******************************************************************\
+
+Module: Util
+
+Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_ARRAY_ELEMENT_FROM_POINTER_H
+#define CPROVER_UTIL_ARRAY_ELEMENT_FROM_POINTER_H
+
+#include "std_expr.h"
+
+/// Generate statement using pointer arithmetic to access the element at the
+/// given index of a pointer array:
+/// `*(pointer + index)`
+/// Arrays are sometimes (always in JBMC) represented as a pointer to their
+/// first element. This is because we cannot know the type of an array of
+/// nondeterministic length, as the length is part of any array type. But we do
+/// know the type of the first element of the array, so we work with that
+/// instead.
+/// \param pointer: pointer to the first element of an array
+/// \param index: index of the element to access
+/// \return expression representing the (\p index)'th element of the array
+dereference_exprt
+array_element_from_pointer(const exprt &pointer, const exprt &index)
+{
+  PRECONDITION(can_cast_type<pointer_typet>(pointer.type()));
+  return dereference_exprt{plus_exprt{pointer, index}};
```

Comment:
While at it, maybe also do `PRECONDITION(index.type().id() == ID_signedbv || index.type().id() == ID_unsignedbv);`? I think those are the only two reasonable options, although maybe we could also work with some other bitvector types.

Reply:
I added the precondition. If we need other types in the future we can always add them later.